src/main/java/spoon/support/adaption/TypeAdaptor.java

```diff
@@ -0,0 +1,598 @@
+		this.hierarchyStart = hierarchyStartReference.getTypeDeclaration();
+		this.initializedWithReference = true;
+	}
+
+	/**
+	 * Creates a new type adaptor using the given method as the start of its hierarchy.
+	 *
+	 * @param hierarchyStart the start of the hierarchy
+	 */
+	public TypeAdaptor(CtMethod<?> hierarchyStart) {
+		this(hierarchyStart.getDeclaringType());
+		this.startMethod = hierarchyStart;
+	}
+
+	/**
+	 * Checks if the context of this type adapter is a subtype of the passed superRef.
+	 *
+	 * @param superRef the super reference to check against
+	 * @return true if the context of this type adapter is a subtype of the passed superRef
+	 * @implNote This implementation behaves the same as {@code isSubtype(hierarchyStart,
+	 * 	superRef)}
+	 * @see #isSubtype(CtType, CtTypeReference)
+	 */
+	@SuppressWarnings("removal")
+	public boolean isSubtypeOf(CtTypeReference<?> superRef) {
+		if (superRef.getFactory().getEnvironment().useOldAndSoonDeprecatedClassContextTypeAdaption()) {
+			return getOldClassTypingContext().isSubtypeOf(superRef);
+		}
+		// We have no declaration so we can't really do any subtype queries
+		if (hierarchyStart == null) {
+			return false;
+		}
+
+		boolean subtype = isSubtype(hierarchyStart, superRef);
+		if (!subtype) {
+			return false;
+		}
+		if (hierarchyStartReference.getActualTypeArguments().isEmpty() && superRef.getActualTypeArguments().isEmpty()) {
+			return true;
+		}
+		return new ClassTypingContext(hierarchyStartReference).isSubtypeOf(superRef);
+	}
+
+	/**
+	 * @return the context of this type adaptor
+	 */
+	public CtType<?> getHierarchyStart() {
+		return hierarchyStart;
+	}
+
+	/**
+	 * Checks whether the base is a subtype of the passed superref. Generic parameters in superRef are
+	 * ignored.
+	 *
+	 * @param base the base type
+	 * @param superRef the potential supertype
+	 * @return true if base extends/implements the super type
+	 */
+	@SuppressWarnings("removal")
+	public static boolean isSubtype(CtType<?> base, CtTypeReference<?> superRef) {
+		if (base.getFactory().getEnvironment().useOldAndSoonDeprecatedClassContextTypeAdaption()) {
+			return new TypeAdaptor(base).isSubtypeOf(superRef);
+		}
+		String superRefFqn = superRef.getTypeErasure().getQualifiedName();
+
+		// Everything inherits from object
+		if (superRef.getQualifiedName().equals("java.lang.Object")) {
+			return true;
+		}
+
+		// Types are subtypes of themselves
+		if (base.getQualifiedName().equals(superRefFqn)) {
+			return true;
+		}
+
+		// Walk up the supertype hierarchy and see if we find the super ref type
+		Queue<CtTypeReference<?>> workQueue = new ArrayDeque<>();
+		workQueue.add(base.getReference());
+
+		while (!workQueue.isEmpty()) {
+			CtTypeReference<?> next = workQueue.poll();
+			if (next.getQualifiedName().equals(superRefFqn)) {
+				return true;
+			}
+
+			if (next.getSuperclass() != null) {
+				workQueue.add(next.getSuperclass());
+			}
+			workQueue.addAll(next.getSuperInterfaces());
+		}
```

Comment:
This seems like a very specific thing. As you want to comment _what_ this does, I'd suggest making it a method of its own.

Reply:
I tried, but coming up with a descriptive name was non trivial. I am not perfectly happy with what I ended up with, but it hopefully works. As always, I am open for suggestions though :)